borsh-derive-internal/src/enum_discriminant_map.rs

```diff
@@ -6,19 +6,26 @@ use syn::{punctuated::Punctuated, token::Comma, Variant};
 
 /// Calculates the discriminant that will be assigned by the compiler.
 /// See: https://doc.rust-lang.org/reference/items/enumerations.html#assigning-discriminant-values
-pub fn discriminant_map(variants: &Punctuated<Variant, Comma>) -> HashMap<Ident, TokenStream> {
+pub fn discriminant_map(
+    variants: &Punctuated<Variant, Comma>,
+) -> (HashMap<Ident, TokenStream>, bool) {
```

Comment:
Instead of changing this function, I would do it on the call sight:

```rust
let has_explicit_discriminants = variants.iter().any(|variant| variant.discriminant.is_some());
```

Reply:
Sorry, it's not quite the same. I need to keep my ugly version.

Reply:
@iho Please, help me understand why it is not quite the same 🙏

Reply:
@frol
The code that you suggested earlier returns true every time it processes a struct or enum, but I want to know if an enum has any explicit discriminant. 
So whatever it has a non-zero value we return true. 
```rust
if !this_discriminant.to_string().starts_with('0') {
            has_discriminant = true;
}
```
Does it make sense? I could be wrong, and tests are not good enough because right now it looks like a legit way to check if an enum has explicit discriminant.

Reply:
@iho [According to `syn` documentation](https://docs.rs/syn/latest/syn/struct.Variant.html#structfield.discriminant) `variant.discriminant` is `Option`, and only set if it is explicitly specified, and if we deconstruct the code we currently have:

```rust
for variant in variants {
        let this_discriminant = variant.discriminant.clone().map_or_else(
            || quote! { #next_discriminant_if_not_specified },
            |(_, e)| quote! { #e },
        );
        if !this_discriminant.to_string().starts_with('0') {
```

and look closer at:

```rust
        if !this_discriminant.to_string().starts_with('0') {
```

it means that:

```rust
for variant in variants {
        let this_discriminant = variant.discriminant.clone().map_or_else(
            || quote! { #next_discriminant_if_not_specified },
            |(_, e)| quote! { #e },  // <-- this is the path to make `this_discriminant` NOT start with '0'
        );
```

which basically means that `variant.discriminant` was `Some`, and thus my proposed oneliner should work exactly the same as it currently works in this function.

Reply:
I don't use the latest version of syn.  I will update the dependency in this branch and check if your code works

Reply:
Your code is correct. Thank you!